File: src/adapters/php/src/LpdfEngine.php
<?php

declare(strict_types=1);

namespace Lpdf;

final class LpdfEngine
{
    /** @var array<string, string> Font name → raw TTF/OTF bytes */
    private array $fonts = [];

    /** @var array<string, string> Image name → raw image bytes (PNG/JPEG/WebP/…) */
    private array $images = [];

    /**
     * Optional RC4-128 encryption config.
     * Keys: user_password (string), owner_password (string), permissions (array).
     * @var array{user_password: string, owner_password: string, permissions: array<string, bool>}|null
     */
    private ?array $encrypt = null;

    public function __construct(
        private readonly string        $licenseKey,
        private readonly RenderOptions $options = new RenderOptions(),
    ) {}

    /**
     * Configure RC4-128 encryption for all subsequent renderPdf() calls.
     * Pass null to clear previously set encryption.
     *
     * @param string                    $userPassword  Open password (empty = no open password).
     * @param string                    $ownerPassword Owner (permissions) password.
     * @param array<string, bool>       $permissions   Flags: print, modify, copy, annotate,
     *                                                 fill_forms, accessibility, assemble, print_hq.
     *                                                 Omitted flags default to true (allowed).
     */
    public function setEncryption(string $userPassword, string $ownerPassword, array $permissions = []): static
    {
        $this->encrypt = [
            'user_password'  => $userPassword,
            'owner_password' => $ownerPassword,
            'permissions'    => $permissions,
        ];
        return $this;
    }

    /**
     * Remove any previously configured encryption.
     */
    public function clearEncryption(): static
    {
        $this->encrypt = null;
        return $this;
    }

    /**
     * Register a custom font asset.
     *
     * The name must match the `ref` attribute of a `<font>` element in the
     * XML `<assets>` block, e.g.:
     *   XML:  <font name="heading" ref="inter-bold"/>
     *   PHP:  $engine->loadFont('inter-bold', file_get_contents('InterBold.ttf'));
     */
    public function loadFont(string $name, string $bytes): static
    {
        $this->fonts[$name] = $bytes;
        return $this;
    }

    /**
     * Register an image asset.
     *
     * The name must match the `name` attribute of an `<image>` element in the
     * XML `<assets>` block, e.g.:
     *   XML:  <image name="logo"/>
     *   PHP:  $engine->loadImage('logo', file_get_contents('logo.png'));
     */
    public function loadImage(string $name, string $bytes): static
    {
        $this->images[$name] = $bytes;
        return $this;
    }

    /**
     * Render an lpdf XML string or LpdfDocument tree and return raw PDF bytes.
     *
     * @param  string|LpdfDocument $input       XML string or a tree built with LpdfKit.
     * @param  RenderOptions|null  $callOptions Per-call overrides merged with constructor options.
     * @throws LpdfRenderException On render or process error.
     */
    public function renderPdf(string|LpdfDocument $input, ?RenderOptions $callOptions = null): string
    {
        if ($input instanceof LpdfDocument) {
            $method   = 'render_tree_pdf';
            $inputStr = json_encode($input, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE);
        } else {
            $method   = 'render_pdf';
            $inputStr = $input;
        }

        $runner = new WasmRunner(
            wasmBinary: $callOptions?->wasmBinary ?? $this->options->wasmBinary ?? self::defaultBinary(),
            wasmRunner: $callOptions?->wasmRunner ?? $this->options->wasmRunner ?? 'wasmtime',
        );

        // Merge fonts: loadFont() calls take precedence, then per-call fontBytes,
        // then constructor-level fontBytes. Per-call wins over constructor on collision.
        $mergedFonts = array_merge(
            $this->options->fontBytes ?? [],
            $callOptions?->fontBytes  ?? [],
            $this->fonts,
        );

        // Auto-load fonts declared via src= that haven't been explicitly provided.
        $fontSrcs = $method === 'render_pdf'
            ? self::xmlFontSrcs($inputStr)
            : self::jsonFontSrcs($inputStr);
        foreach ($fontSrcs as $key => $path) {
            if (!array_key_exists($key, $mergedFonts) && is_readable($path)) {
                $bytes = file_get_contents($path);
                if ($bytes !== false) {
                    $mergedFonts[$key] = $bytes;
                }
            }
        }

        // Merge images: same precedence order as fonts.
        $mergedImages = array_merge(
            $this->options->imageBytes ?? [],
            $callOptions?->imageBytes  ?? [],
            $this->images,
        );

        // Auto-load images declared via src= that haven't been explicitly provided.
        $imageSrcs = $method === 'render_pdf'
            ? self::xmlImageSrcs($inputStr)
            : self::jsonImageSrcs($inputStr);
        foreach ($imageSrcs as $key => $path) {
            if (!array_key_exists($key, $mergedImages) && is_readable($path)) {
                $bytes = file_get_contents($path);
                if ($bytes !== false) {
                    $mergedImages[$key] = $bytes;
                }
            }
        }

        $payload = [
            'method' => $method,
            'key'    => $this->licenseKey,
            'input'  => $inputStr,
        ];

        if ($mergedFonts !== []) {
            $payload['fonts'] = array_map('base64_encode', $mergedFonts);
        }

        if ($mergedImages !== []) {
            $payload['images'] = array_map('base64_encode', $mergedImages);
        }

        $createdOn = $callOptions?->createdOn ?? $this->options->createdOn;
        if ($createdOn !== null) {
            $payload['created_on'] = $createdOn;
        }

        if ($this->encrypt !== null) {
            $payload['encrypt'] = $this->encrypt;
        }

        $response = $runner->invoke($payload);

        if (!isset($response['pdf'])) {
            throw new LpdfRenderException('Unexpected response from WASI process.');
        }

        $bytes = base64_decode($response['pdf'], strict: true);
        if ($bytes === false) {
            throw new LpdfRenderException('Failed to decode base64 PDF response.');
        }

        return $bytes;
    }

    /**
     * Convert an LpdfDocument tree (built with LpdfKit) to an lpdf XML string.
     *
     * The conversion is performed by the Rust core running as a WASI subprocess,
     * so the output is identical to the XML produced by the other adapters.
     *
     * @param  LpdfDocument       $doc         The document tree to serialise.
     * @param  string|null        $wasmBinary  Path to lpdf-wasi.wasm (defaults to bundled binary).
     * @param  string             $wasmRunner  WASI runtime executable (default: wasmtime).
     * @throws LpdfRenderException On process or serialisation error.
     */
    public static function kitToXml(
        LpdfDocument $doc,
        ?string $wasmBinary = null,
        string $wasmRunner = 'wasmtime',
    ): string {
        $runner = new WasmRunner(
            wasmBinary: $wasmBinary ?? self::defaultBinary(),
            wasmRunner: $wasmRunner,
        );
        $payload  = [
            'method' => 'kit_to_xml',
            'input'  => json_encode($doc, JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE),
        ];
        $response = $runner->invoke($payload);
        if (!isset($response['xml'])) {
            throw new LpdfRenderException('Unexpected response from WASI process (kit_to_xml).');
        }
        return $response['xml'];
    }

    /** @return array<string,string> Font ref??name → file path from `<font name="…" src="…">` tags. */
    private static function xmlFontSrcs(string $xml): array
    {
        $srcs = [];
        preg_match_all('/<font\s[^>]*>/', $xml, $m);
        foreach ($m[0] as $tag) {
            if (preg_match('/\bname="([^"]*)"/', $tag, $nm) &&
                preg_match('/\bsrc="([^"]*)"/', $tag, $src)) {
                $key = preg_match('/\bref="([^"]*)"/', $tag, $ref) ? $ref[1] : $nm[1];
                $srcs[$key] = $src[1];
            }
        }
        return $srcs;
    }

    /** @return array<string,string> Image ref??name → file path from `<image name="…" src="…">` tags. */
    private static function xmlImageSrcs(string $xml): array
    {
        $srcs = [];
        preg_match_all('/<image\s[^>]*>/', $xml, $m);
        foreach ($m[0] as $tag) {
            if (preg_match('/\bname="([^"]*)"/', $tag, $nm) &&
                preg_match('/\bsrc="([^"]*)"/', $tag, $src)) {
                $key = preg_match('/\bref="([^"]*)"/', $tag, $ref) ? $ref[1] : $nm[1];
                $srcs[$key] = $src[1];
            }
        }
        return $srcs;
    }

    /** @return array<string,string> Font ref??name → file path from a serialised tree's `tokens.fonts`. */
    private static function jsonFontSrcs(string $json): array
    {
        $srcs = [];
        $doc  = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
        foreach ($doc['attrs']['tokens']['fonts'] ?? [] as $name => $def) {
            if (isset($def['src'])) {
                $key = $def['ref'] ?? $name;
                $srcs[$key] = $def['src'];
            }
        }
        return $srcs;
    }

    /** @return array<string,string> Image ref??name → file path from a serialised tree's `tokens.images`. */
    private static function jsonImageSrcs(string $json): array
    {
        $srcs = [];
        $doc  = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
        foreach ($doc['attrs']['tokens']['images'] ?? [] as $name => $def) {
            if (isset($def['src'])) {
                $key = $def['ref'] ?? $name;
                $srcs[$key] = $def['src'];
            }
        }
        return $srcs;
    }

    private static function defaultBinary(): string
    {
        return \dirname(__DIR__) . '/resources/lpdf-wasi.wasm';
    }
}

File: src/adapters/dotnet/src/LpdfEngine.cs
using System.Text.Json;

namespace Lpdf;

/// <summary>
/// Options shared across multiple <see cref="LpdfEngine.RenderPdf(string, RenderOptions?)"/> calls.
/// </summary>
public sealed class RenderOptions
{
    /// <summary>
    /// Pre-loaded font bytes for custom fonts referenced via <c>fonts src="…"</c>.
    /// Keys are the font token names used in the document; values are raw TTF/OTF bytes.
    /// </summary>
    public IReadOnlyDictionary<string, byte[]>? FontBytes { get; init; }

    /// <summary>
    /// Pre-loaded image bytes for images referenced via <c>&lt;img name="…"&gt;</c>.
    /// Keys are the image names declared in <c>&lt;assets&gt;</c>; values are raw PNG/JPEG bytes.
    /// </summary>
    public IReadOnlyDictionary<string, byte[]>? ImageBytes { get; init; }

    /// <summary>
    /// File-read callback for resolving font <c>src</c> paths at render time.
    /// On the server this can be set to <c>System.IO.File.ReadAllBytes</c>.
    /// In sandboxed environments supply all bytes via <see cref="FontBytes"/>.
    /// </summary>
    public Func<string, byte[]>? SrcFallback { get; init; }
}

/// <summary>
/// Stateful lpdf renderer.
/// Construct once with a license key; call <see cref="RenderPdf(string, RenderOptions?)"/>
/// or <see cref="RenderPdf(LpdfDocument, RenderOptions?)"/> as many times as needed.
/// </summary>
public sealed class LpdfEngine : IDisposable
{
    private readonly string         _licenseKey;
    private readonly RenderOptions  _opts;
    private readonly WasmRunner     _wasm;
    private readonly Dictionary<string, byte[]> _fonts  = new();
    private readonly Dictionary<string, byte[]> _images = new();
    private          bool           _disposed;

    /// <param name="licenseKey">
    ///   Your lpdf license key. Pass an empty string to render in evaluation
    ///   mode (produces a visible watermark).
    /// </param>
    /// <param name="options">Shared render options applied to every call.</param>
    public LpdfEngine(string licenseKey, RenderOptions? options = null)
    {
        _licenseKey = licenseKey ?? throw new ArgumentNullException(nameof(licenseKey));
        _opts       = options ?? new RenderOptions();
        _wasm       = new WasmRunner();
    }

    // ── Public API ────────────────────────────────────────────────────────────

    /// <summary>
    /// Register raw TTF/OTF bytes for a font name referenced via <c>fonts src="…"</c>.
    /// Call before <see cref="RenderPdf(string, RenderOptions?)"/>. Returns <c>this</c>.
    /// </summary>
    public LpdfEngine LoadFont(string name, byte[] bytes)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(bytes);
        _fonts[name] = bytes;
        return this;
    }

    /// <summary>
    /// Register raw PNG or JPEG bytes for an image name referenced via <c>&lt;img name="…"&gt;</c>.
    /// Call before <see cref="RenderPdf(string, RenderOptions?)"/>. Returns <c>this</c>.
    /// </summary>
    public LpdfEngine LoadImage(string name, byte[] bytes)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(name);
        ArgumentNullException.ThrowIfNull(bytes);
        _images[name] = bytes;
        return this;
    }

    /// <summary>Render an lpdf XML string to PDF bytes.</summary>
    /// <param name="xml">Full lpdf XML document string.</param>
    /// <param name="callOptions">Per-call overrides merged with the constructor options.</param>
    public Task<byte[]> RenderPdf(string xml, RenderOptions? callOptions = null)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(xml);
        var merged = Merge(callOptions);
        return Task.FromResult(
            _wasm.RenderPdf(xml, _licenseKey, merged.FontBytes, merged.ImageBytes, merged.SrcFallback));
    }

    /// <summary>Render an <see cref="LpdfDocument"/> tree (built with <see cref="LpdfKit"/>) to PDF bytes.</summary>
    /// <param name="document">Document tree produced by <c>LpdfKit.Document(…)</c>.</param>
    /// <param name="callOptions">Per-call overrides merged with the constructor options.</param>
    public Task<byte[]> RenderPdf(LpdfDocument document, RenderOptions? callOptions = null)
    {
        ThrowIfDisposed();
        ArgumentNullException.ThrowIfNull(document);
        var merged = Merge(callOptions);
        var json   = JsonSerializer.Serialize(document, LpdfDocumentJson.Options);
        return Task.FromResult(
            _wasm.RenderTreePdf(json, _licenseKey, merged.FontBytes, merged.ImageBytes, merged.SrcFallback));
    }

    // ── Internals ─────────────────────────────────────────────────────────────

    private RenderOptions Merge(RenderOptions? call)
    {
        if (call is null && _fonts.Count == 0 && _images.Count == 0) return _opts;
        return new RenderOptions
        {
            FontBytes   = MergeDicts(MergeDicts(_opts.FontBytes, call?.FontBytes), _fonts.Count > 0 ? _fonts : null),
            ImageBytes  = MergeDicts(MergeDicts(_opts.ImageBytes, call?.ImageBytes), _images.Count > 0 ? _images : null),
            SrcFallback = call?.SrcFallback ?? _opts.SrcFallback,
        };
    }

    private static IReadOnlyDictionary<string, byte[]>? MergeDicts(
        IReadOnlyDictionary<string, byte[]>? a,
        IReadOnlyDictionary<string, byte[]>? b)
    {
        if (a is null) return b;
        if (b is null) return a;
        var merged = new Dictionary<string, byte[]>(a);
        foreach (var (k, v) in b) merged[k] = v;
        return merged;
    }

    private void ThrowIfDisposed()
        => ObjectDisposedException.ThrowIf(_disposed, this);

    /// <summary>Releases the WASM runtime resources held by this engine.</summary>
    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _wasm.Dispose();
    }
}

/// <summary>Thrown when the lpdf engine returns a layout or parse error.</summary>
public sealed class LpdfRenderException(string message) : Exception(message);

